rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+//          modulus: u64,
+//          p: &[u8],
+//          q: &[u8],
+//              ...
+//        }
+
+///
+// Rsa functions. For further documentation please refer to symcrypt.h
+// TODO rest of the documentation
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+/// 2^16 + 1
+const DEFAULT_PUBLIC_EXPONENT: u64 = 0x10001;
+
+/// Rsa Public and Private Key State.
+///
+/// [`RsaKeyPair`] stores a private and public Rsa key.
+pub struct RsaKeyPair {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+}
+
+/// Rsa Public Key State
+///
+/// [`RsaPublicKey`] stores an Rsa public key.
+pub struct RsaPublicKey {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `RsaKeyUsage` will indicate if the [`RsaKeyPair`] or [`RsaPublicKey`] will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`].
+/// 
+/// [`RsaKeyUsage::Encrypt`] will have both [`RsaKeyUsage::Sign`] and [`RsaKeyUsage::Encrypt`] flags set.
+pub enum RsaKeyUsage {
+    Sign,
+    Encrypt,
+}
+
+// !Review: Should we add the no fips / minimal validation flags? The whole purpose is to eventually get to fips, adding these extra flags might not be needed?
+// #define SYMCRYPT_FLAG_KEY_NO_FIPS               (0x100)
+// #define SYMCRYPT_FLAG_KEY_MINIMAL_VALIDATION    (0x200)
+// #define SYMCRYPT_FLAG_RSAKEY_SIGN       (0x1000) 
+// #define SYMCRYPT_FLAG_RSAKEY_ENCRYPT    (0x2000)
+impl RsaKeyUsage {
+    pub fn to_flag(&self) -> symcrypt_sys::UINT32 {
+        match self {
+            RsaKeyUsage::Sign => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN,
+            RsaKeyUsage::Encrypt => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT | symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN ,
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct RsaKeyPairBlob {
+    modulus: Vec<u8>,
+    pub_exp: Vec<u8>,
+    p: Vec<u8>,
+    q: Vec<u8>,
+    d_p: Vec<u8>,
+    d_q: Vec<u8>,
+    crt_coefficient: Vec<u8>,
+    private_exp: Vec<u8>,
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct RsaPublicKeyBlob {
+    modulus: Vec<u8>,
+    pub_exp: Vec<u8>,
+}
+
+impl RsaKeyPair {
+    /// `generate_new()` generates a random Rsa key based on the provided parameters.
+    ///
+    /// `n_bits_mod` represents the desired bit length of the Rsa Key.
+    ///
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    ///  
+    /// `key_usage` will indicate if this key will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`]
+    pub fn generate_new(
+        n_bits_mod: u32,
+        pub_exp: Option<&[u8]>,
+        key_usage: RsaKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let rsa_key = allocate_rsa(2, n_bits_mod)?; // !Review: @Phil has mentioned that there is no scenario that we'd want to ONLY generate a public key, @Crypto folks is this the case?
```

Comment:
>!Review: @Phil has mentioned that there is no scenario that we'd want to ONLY generate a public key, @Crypto folks is this the case? [](http://example.com/codeflow?start=55&length=132)

Yeah. I think SymCrypt does support this, but generating a public key without a corresponding private key doesn't make sense. It would be useless. You can import a public key only, but in that case it was generated by someone else and you're just using it for encryption or verification.